tawazi/_dag/helpers.py

```diff
@@ -0,0 +1,298 @@
+    return bool(xns_dict[xn.active.id].result)
+
+
+def copy_non_setup_xns(x_nodes: Dict[str, ExecNode]) -> Dict[str, ExecNode]:
+    """Deep copy all ExecNodes except setup ExecNodes because they are shared throughout the DAG instance.
+
+    Args:
+        x_nodes: Dict[str, ExecNode] x_nodes to be deep copied
+
+    Returns:
+        Dict[str, ExecNode] copy of x_nodes
+    """
+    # TODO: separate setup xnodes and non setup xndoes.
+    #  maybe use copy instead of deepcopy for the non setup xnodes!? I think this is a bad idea it won't work
+    x_nodes_copy = {}
+    for id_, x_nd in x_nodes.items():
+        # if execnode is a setup node, it shouldn't be copied
+        if x_nd.setup:
+            x_nodes_copy[id_] = x_nd
+        else:
+            # no need to deepcopy. we only need to know if self.result is NoVal or not (TODO: fix this COmment)
+            x_nodes_copy[id_] = copy(x_nd)
+    return x_nodes_copy
+
+
+#####################
+# scheduler Functions
+#####################
+
+
+def get_num_running_threads(_futures: Dict[Identifier, "Future[Any]"]) -> int:
+    # use not future.done() because there is no guarantee that Thread pool will directly execute
+    # the submitted thread
+    return sum([not future.done() for future in _futures.values()])
+
+
+def get_highest_priority_nodes(nodes: List[ExecNode]) -> List[ExecNode]:
+    highest_priority = max(node.priority for node in nodes)
+    return [node for node in nodes if node.priority == highest_priority]
+
+
+def get_next_timeout(
+    futures_launch_time: Dict[Identifier, float],
+    futures: Dict[Identifier, "Future[Any]"],
+    xns_dict: Dict[Identifier, ExecNode],
+) -> Optional[float]:
+    """Get the timeout to wait before an ExecNode issuance date arrives.
+
+    Args:
+        futures_launch_time (Dict[Identifier, float]): The launch time of the futures (submission time).
+        futures (Dict[Identifier, &quot;Future[Any]&quot;]): The futures that are submitted (running, scheduled & done).
+        xns_dict (Dict[Identifier, ExecNode]): The ExecNodes whose Futures are submitted.
+
+    Returns:
+        Optional[float]: The next timeout to use in wait.
+    """
+    # start next_timeout as infinity.
+    # 1st timeout encountered will reduce it to a finite number.
+    next_timeout = float("inf")
+    for id_, fut in futures.items():
+        # if future is not done, then it should be running (This is not guaranteed in ThreadPoolExecutor)
+        if not fut.done():
+            launch_time = futures_launch_time[id_]
+            timeout = xns_dict[id_].timeout
+            logger.debug(
+                f"next_timeout: {next_timeout}, launch_time: {launch_time}, timeout: {timeout}"
+            )
+
+            if timeout is not None:
+                next_timeout = min(next_timeout, launch_time + timeout)
+
+    if next_timeout == float("inf"):
+        return None
+    return next_timeout - monotonic()
+
+
+def raise_timeout_error_conditionally(
```

Comment:
rather `raise_conditional_timeout_error` ?

Reply:
I agree that the function name is not clear..
Gihubcopilot proposed those two names:
```
Based on the functionality of the function, a possible name for it could be check_timeout_and_raise_error() or raise_timeout_error_if_needed(). These names reflect the fact that the function checks if a timeout has been reached and raises a TawaziTimeoutError exception if needed.
```

What do you thing ? I like `check_timeout_and_raise_error`

Reply:
`check_timeout` might also be a good pick ? It's simpler and still clear don't you think ?

Reply:
yes this is good

Reply:
done